File: sdk-go/httrace.go
// Package httrace provides an HTTP middleware for capturing real production
// traffic and sending it to the Httrace API for automatic test generation.
//
// Works with the standard net/http library — and therefore with chi, gorilla/mux,
// gin (via gin.WrapH), echo, and any other Go HTTP framework.
//
// Usage (net/http / chi):
//
//	import "github.com/httrace-io/httrace-go"
//
//	cfg := httrace.Config{APIKey: "ht_...", Service: "orders-api"}
//	http.Handle("/", httrace.Middleware(cfg)(yourHandler))
//
// Usage (gin):
//
//	r.Use(gin.WrapH(httrace.Middleware(cfg)(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {}))))
package httrace

import (
	"bytes"
	"context"
	"encoding/json"
	"io"
	"math/rand"
	"net/http"
	"net/url"
	"regexp"
	"strings"
	"sync"
	"time"
)

const defaultEndpoint = "https://api.httrace.com/v1/captures"

var sensitiveHeaders = map[string]bool{
	"authorization":        true,
	"cookie":              true,
	"set-cookie":          true,
	"x-api-key":           true,
	"x-auth-token":        true,
	"proxy-authorization": true,
}

var binaryContentTypes = []string{
	"application/octet-stream", "image/", "audio/", "video/",
	"application/gzip", "application/zip", "application/pdf", "multipart/",
}

var sensitiveKeys = []string{"password", "secret", "token", "ssn", "credit_card", "card_number", "cvv"}

// outgoingCallKey is the context key for per-request outgoing call capture.
type outgoingCallKey struct{}

// OutgoingCall represents a single outgoing dependency call captured during a request.
type OutgoingCall struct {
	Type           string      `json:"type"`
	Method         string      `json:"method,omitempty"`
	URL            string      `json:"url,omitempty"`
	RequestBody    interface{} `json:"request_body,omitempty"`
	ResponseStatus int         `json:"response_status,omitempty"`
	ResponseBody   interface{} `json:"response_body,omitempty"`
	LatencyMs      float64     `json:"latency_ms"`
}

var sensitiveURLParamRe = regexp.MustCompile(`(?i)api[-_]?key|apikey|token|secret|auth|password|passwd|credential|access[-_]?token`)

func sanitizeOutgoingURL(rawURL string) string {
	parsed, err := url.Parse(rawURL)
	if err != nil {
		return rawURL
	}
	q := parsed.Query()
	for k := range q {
		if sensitiveURLParamRe.MatchString(k) {
			q.Set(k, "<REDACTED>")
		}
	}
	parsed.RawQuery = q.Encode()
	return parsed.String()
}

// RecordingTransport is an http.RoundTripper that records outgoing HTTP calls
// into the per-request context when CaptureOutgoing is enabled.
type RecordingTransport struct {
	Base http.RoundTripper
}

func (t *RecordingTransport) RoundTrip(req *http.Request) (*http.Response, error) {
	calls, _ := req.Context().Value(outgoingCallKey{}).(*[]OutgoingCall)

	base := t.Base
	if base == nil {
		base = http.DefaultTransport
	}

	if calls == nil {
		return base.RoundTrip(req)
	}

	t0 := time.Now()
	resp, err := base.RoundTrip(req)
	latency := time.Since(t0).Seconds() * 1000

	if err == nil {
		var body interface{}
		ct := resp.Header.Get("Content-Type")
		if strings.Contains(ct, "application/json") {
			raw, readErr := io.ReadAll(resp.Body)
			resp.Body.Close()
			if readErr == nil {
				_ = json.Unmarshal(raw, &body)
				resp.Body = io.NopCloser(bytes.NewReader(raw))
			}
		}
		*calls = append(*calls, OutgoingCall{
			Type:           "http",
			Method:         req.Method,
			URL:            sanitizeOutgoingURL(req.URL.String()),
			ResponseStatus: resp.StatusCode,
			ResponseBody:   body,
			LatencyMs:      latency,
		})
	}

	return resp, err
}

// ClientFromContext returns an *http.Client whose transport records outgoing calls
// into the per-request context. Use this client for any outgoing HTTP calls
// you want to capture when CaptureOutgoing is enabled in the middleware.
//
//	client := httrace.ClientFromContext(r.Context())
//	resp, err := client.Get("https://api.stripe.com/...")
func ClientFromContext(ctx context.Context) *http.Client {
	return &http.Client{
		Transport: &RecordingTransport{Base: http.DefaultTransport},
	}
}

// contextWithOutgoingCapture returns a context that records outgoing calls.
func contextWithOutgoingCapture(ctx context.Context) (context.Context, *[]OutgoingCall) {
	calls := make([]OutgoingCall, 0)
	return context.WithValue(ctx, outgoingCallKey{}, &calls), &calls
}

// Config holds configuration for the Httrace middleware.
type Config struct {
	// APIKey is your Httrace API key — required.
	APIKey string

	// Service labels this service in the Httrace dashboard (default: "default").
	Service string

	// SampleRate is the fraction of requests to capture (0.0–1.0, default: 0.1).
	SampleRate float64

	// ExcludePaths is a list of URL paths to skip (default: /health, /metrics, /favicon.ico).
	ExcludePaths []string

	// Endpoint overrides the default Httrace API endpoint (useful for self-hosted).
	Endpoint string

	// CaptureOutgoing enables recording of outgoing HTTP calls made during a request.
	// Use ClientFromContext(r.Context()) to obtain an instrumented *http.Client.
	CaptureOutgoing bool
}

// Middleware returns an http.Handler middleware that captures traffic.
func Middleware(cfg Config) func(http.Handler) http.Handler {
	if cfg.Service == "" {
		cfg.Service = "default"
	}
	if cfg.SampleRate == 0 {
		cfg.SampleRate = 0.1
	}
	if cfg.Endpoint == "" {
		cfg.Endpoint = defaultEndpoint
	}

	exclude := map[string]bool{
		"/health": true, "/metrics": true, "/favicon.ico": true,
	}
	for _, p := range cfg.ExcludePaths {
		exclude[p] = true
	}

	c := newClient(cfg.APIKey, cfg.Endpoint)

	captureOutgoing := cfg.CaptureOutgoing

	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			if exclude[r.URL.Path] || rand.Float64() >= cfg.SampleRate {
				next.ServeHTTP(w, r)
				return
			}

			// Buffer request body so both us and the handler can read it
			reqBody, _ := io.ReadAll(r.Body)
			r.Body = io.NopCloser(bytes.NewReader(reqBody))

			// Optionally activate outgoing call capture for this request
			var outgoingCalls *[]OutgoingCall
			if captureOutgoing {
				ctx, calls := contextWithOutgoingCapture(r.Context())
				r = r.WithContext(ctx)
				outgoingCalls = calls
			}

			rw := &responseRecorder{ResponseWriter: w, statusCode: 200}
			tStart := time.Now()
			next.ServeHTTP(rw, r)
			latency := time.Since(tStart).Seconds() * 1000

			go record(c, cfg.Service, r, reqBody, rw, latency, outgoingCalls)
		})
	}
}

func record(c *client, service string, r *http.Request, reqBody []byte, rw *responseRecorder, latency float64, outgoingCalls *[]OutgoingCall) {
	defer func() { recover() }()

	query := make(map[string]string)
	for k, v := range r.URL.Query() {
		if len(v) > 0 {
			query[k] = v[0]
		}
	}

	var outgoing []OutgoingCall
	if outgoingCalls != nil {
		outgoing = *outgoingCalls
	}

	interaction := map[string]interface{}{
		"service":    service,
		"session_id": firstNonEmpty(r.Header.Get("X-Session-Id"), r.Header.Get("X-Request-Id")),
		"request": map[string]interface{}{
			"method":       r.Method,
			"path":         r.URL.Path,
			"query_params": query,
			"headers":      filterHeaders(r.Header),
			"body":         sanitize(parseBody(reqBody, r.Header.Get("Content-Type"))),
			"timestamp":    float64(time.Now().UnixNano()) / 1e9,
		},
		"response": map[string]interface{}{
			"status_code": rw.statusCode,
			"headers":     map[string]string{},
			"body":        sanitize(parseBody(rw.body.Bytes(), rw.Header().Get("Content-Type"))),
			"latency_ms":  latency,
		},
		"outgoing_calls": outgoing,
	}

	c.enqueue(interaction)
}

// ── responseRecorder captures status code + body ───────────────────────────

type responseRecorder struct {
	http.ResponseWriter
	statusCode int
	body       bytes.Buffer
}

func (rr *responseRecorder) WriteHeader(code int) {
	rr.statusCode = code
	rr.ResponseWriter.WriteHeader(code)
}

func (rr *responseRecorder) Write(b []byte) (int, error) {
	rr.body.Write(b)
	return rr.ResponseWriter.Write(b)
}

// ── Helpers ────────────────────────────────────────────────────────────────

func filterHeaders(h http.Header) map[string]string {
	out := make(map[string]string)
	for k, v := range h {
		if !sensitiveHeaders[strings.ToLower(k)] && len(v) > 0 {
			out[k] = v[0]
		}
	}
	return out
}

func parseBody(raw []byte, ct string) interface{} {
	if len(raw) == 0 {
		return nil
	}
	for _, t := range binaryContentTypes {
		if strings.Contains(ct, t) {
			return nil
		}
	}
	if strings.Contains(ct, "application/json") {
		var v interface{}
		if json.Unmarshal(raw, &v) == nil {
			return v
		}
	}
	return string(raw)
}

func sanitize(val interface{}) interface{} {
	switch v := val.(type) {
	case string:
		return v // simple strings pass through (PII patterns handled server-side)
	case map[string]interface{}:
		out := make(map[string]interface{}, len(v))
		for k, vv := range v {
			lk := strings.ToLower(k)
			redact := false
			for _, s := range sensitiveKeys {
				if strings.Contains(lk, s) {
					redact = true
					break
				}
			}
			if redact {
				out[k] = "[REDACTED]"
			} else {
				out[k] = sanitize(vv)
			}
		}
		return out
	case []interface{}:
		out := make([]interface{}, len(v))
		for i, vv := range v {
			out[i] = sanitize(vv)
		}
		return out
	}
	return val
}

func firstNonEmpty(vals ...string) string {
	for _, v := range vals {
		if v != "" {
			return v
		}
	}
	return ""
}

// ── Background client ──────────────────────────────────────────────────────

type client struct {
	apiKey   string
	endpoint string
	mu       sync.Mutex
	queue    []interface{}
}

func newClient(apiKey, endpoint string) *client {
	c := &client{apiKey: apiKey, endpoint: endpoint}
	go c.worker()
	return c
}

func (c *client) enqueue(item interface{}) {
	c.mu.Lock()
	defer c.mu.Unlock()
	if len(c.queue) < 10_000 {
		c.queue = append(c.queue, item)
	}
}

func (c *client) worker() {
	ticker := time.NewTicker(2 * time.Second)
	defer ticker.Stop()
	for range ticker.C {
		c.flush()
	}
}

func (c *client) flush() {
	c.mu.Lock()
	if len(c.queue) == 0 {
		c.mu.Unlock()
		return
	}
	batch := c.queue
	c.queue = nil
	c.mu.Unlock()

	body, err := json.Marshal(map[string]interface{}{"captures": batch})
	if err != nil {
		return
	}

	req, err := http.NewRequest(http.MethodPost, c.endpoint, bytes.NewReader(body))
	if err != nil {
		return
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("X-Api-Key", c.apiKey)

	httpClient := &http.Client{Timeout: 5 * time.Second}
	resp, err := httpClient.Do(req)
	if err == nil {
		resp.Body.Close()
	}
}

File: sdk-ruby/lib/httrace.rb
require 'json'
require 'net/http'
require 'uri'
require 'cgi'
require 'set'
require 'thread'
require 'stringio'

module Httrace
  VERSION = '0.1.0'
  DEFAULT_ENDPOINT = 'https://api.httrace.com/v1/captures'

  SENSITIVE_URL_PARAM_RE = /api[-_]?key|apikey|token|secret|auth|password|passwd|credential|access[-_]?token/i

  # ── Outgoing call interceptor for Net::HTTP ─────────────────────────────────

  module NetHTTPInterceptor
    def request(req, body = nil, &block)
      calls = Thread.current[:httrace_outgoing]
      return super unless calls

      t0 = Process.clock_gettime(Process::CLOCK_MONOTONIC)
      response = super
      latency_ms = (Process.clock_gettime(Process::CLOCK_MONOTONIC) - t0) * 1000.0

      begin
        raw_url = "#{use_ssl? ? 'https' : 'http'}://#{address}:#{port}#{req.path}"
        clean_url = Httrace.sanitize_outgoing_url(raw_url)

        body_val = nil
        ct = response['content-type'] || ''
        if ct.include?('application/json')
          begin
            body_val = JSON.parse(response.body)
          rescue
          end
        end

        calls << {
          type:            'http',
          method:          req.method.upcase,
          url:             clean_url,
          response_status: response.code.to_i,
          response_body:   body_val,
          latency_ms:      latency_ms.round(2),
        }
      rescue
      end

      response
    end
  end

  def self.sanitize_outgoing_url(raw_url)
    parsed = URI.parse(raw_url)
    if parsed.query
      qs = URI.decode_www_form(parsed.query).map do |k, v|
        [k, SENSITIVE_URL_PARAM_RE.match?(k) ? '<REDACTED>' : v]
      end
      parsed.query = URI.encode_www_form(qs)
    end
    parsed.to_s
  rescue
    raw_url
  end

  SENSITIVE_HEADERS = Set.new(%w[
    authorization cookie set-cookie x-api-key x-auth-token proxy-authorization
  ])

  BINARY_CONTENT_TYPES = %w[
    application/octet-stream image/ audio/ video/
    application/gzip application/zip application/pdf multipart/
  ]

  SENSITIVE_KEYS = %w[password secret token ssn credit_card card_number cvv]

  # ── CaptureMiddleware ──────────────────────────────────────────────────────
  #
  # Rack middleware — works with Rails, Sinatra, Grape, and any Rack-compatible app.
  #
  # Rails (config/application.rb):
  #   config.middleware.use Httrace::CaptureMiddleware, api_key: 'ht_...'
  #
  # Sinatra:
  #   use Httrace::CaptureMiddleware, api_key: 'ht_...'
  #
  class CaptureMiddleware
    # @param app             [#call]  the Rack app to wrap
    # @param api_key         [String] your Httrace API key
    # @param service         [String] service label shown in the dashboard
    # @param sample_rate     [Float]  fraction of requests to capture (0.0–1.0)
    # @param exclude_paths   [Array]  paths to skip
    # @param endpoint        [String] override API endpoint (for self-hosted)
    # @param capture_outgoing[Boolean] capture outgoing Net::HTTP + ActiveRecord SQL calls
    def initialize(app, api_key:, service: 'default', sample_rate: 0.1,
                   exclude_paths: nil, endpoint: nil, capture_outgoing: false)
      @app              = app
      @service          = service
      @sample_rate      = sample_rate.to_f
      @exclude          = Set.new(exclude_paths || %w[/health /metrics /favicon.ico])
      @client           = Client.new(api_key, endpoint || DEFAULT_ENDPOINT)
      @capture_outgoing = capture_outgoing

      if capture_outgoing
        Net::HTTP.prepend(Httrace::NetHTTPInterceptor)
      end
    end

    def call(env)
      path = env['PATH_INFO'] || '/'

      if @exclude.include?(path) || rand >= @sample_rate
        return @app.call(env)
      end

      # Buffer request body so we can read it AND so the app can read it too
      raw_input = env['rack.input'] || StringIO.new
      req_body  = raw_input.read
      env['rack.input'] = StringIO.new(req_body)

      # Activate outgoing call capture for this request thread
      if @capture_outgoing
        Thread.current[:httrace_outgoing] = []
      end

      t_start = Process.clock_gettime(Process::CLOCK_MONOTONIC)
      status, headers, body_iter = @app.call(env)
      latency_ms = (Process.clock_gettime(Process::CLOCK_MONOTONIC) - t_start) * 1000.0

      # Collect outgoing calls before clearing the thread-local
      outgoing_calls = @capture_outgoing ? (Thread.current[:httrace_outgoing] || []).dup : []
      Thread.current[:httrace_outgoing] = nil if @capture_outgoing

      # Collect response body without consuming it (Rack contract: body is enumerable)
      resp_parts = []
      body_iter.each { |chunk| resp_parts << chunk }
      resp_body_str = resp_parts.join

      Thread.new do
        record(env, req_body, status, headers, resp_body_str, latency_ms, outgoing_calls)
      rescue
        # Never crash
      end

      [status, headers, [resp_body_str]]
    end

    private

    def record(env, req_body, status, resp_headers, resp_body, latency_ms, outgoing_calls = [])
      req_ct   = env['CONTENT_TYPE'] || ''
      resp_ct  = resp_headers['Content-Type'] || resp_headers['content-type'] || ''

      query_params = parse_query(env['QUERY_STRING'])

      interaction = {
        service:        @service,
        session_id:     env['HTTP_X_SESSION_ID'] || env['HTTP_X_REQUEST_ID'],
        outgoing_calls: outgoing_calls,
        request: {
          method:       env['REQUEST_METHOD'],
          path:         env['PATH_INFO'],
          query_params: query_params,
          headers:      filter_headers(env),
          body:         sanitize(parse_body(req_body, req_ct)),
          timestamp:    Time.now.to_f,
        },
        response: {
          status_code: status.to_i,
          headers:     {},
          body:        sanitize(parse_body(resp_body, resp_ct)),
          latency_ms:  latency_ms.round(2),
        },
      }

      @client.enqueue(interaction)
    end

    def filter_headers(env)
      env.each_with_object({}) do |(key, val), h|
        next unless key.start_with?('HTTP_')
        name = key[5..].downcase.tr('_', '-')
        h[name] = val unless SENSITIVE_HEADERS.include?(name)
      end
    end

    def parse_body(raw, content_type)
      return nil if raw.nil? || raw.empty?
      return nil if BINARY_CONTENT_TYPES.any? { |t| content_type.include?(t) }

      if content_type.include?('application/json')
        begin
          return JSON.parse(raw)
        rescue JSON::ParserError
          # fall through
        end
      end

      raw.encode('UTF-8', invalid: :replace, undef: :replace)
    end

    def parse_query(qs)
      return {} if qs.nil? || qs.empty?
      qs.split('&').each_with_object({}) do |pair, h|
        k, v = pair.split('=', 2)
        h[CGI.unescape(k.to_s)] = CGI.unescape(v.to_s) if k
      end
    end

    def sanitize(val)
      case val
      when Hash
        val.each_with_object({}) do |(k, v), h|
          lk = k.to_s.downcase
          h[k] = SENSITIVE_KEYS.any? { |s| lk.include?(s) } ? '[REDACTED]' : sanitize(v)
        end
      when Array
        val.map { |v| sanitize(v) }
      when String
        val
          .gsub(/\b[\w.+\-]+@[\w.\-]+\.\w{2,}\b/, '[EMAIL]')
          .gsub(/\b(?:\d[ \-]?){13,16}\b/, '[CARD]')
      else
        val
      end
    end
  end

  # ── Client ─────────────────────────────────────────────────────────────────

  class Client
    BATCH_SIZE = 50
    MAX_QUEUE  = 10_000

    def initialize(api_key, endpoint)
      @api_key  = api_key
      @uri      = URI.parse(endpoint)
      @queue    = []
      @mu       = Mutex.new
      @worker   = Thread.new { run_worker }
      @worker.abort_on_exception = false
    end

    def enqueue(interaction)
      @mu.synchronize do
        @queue << interaction if @queue.size < MAX_QUEUE
      end
    end

    private

    def run_worker
      loop do
        sleep 2
        flush
      end
    rescue
      # ignore
    end

    def flush
      batch = @mu.synchronize do
        return if @queue.empty?
        b = @queue.dup
        @queue.clear
        b
      end

      return if batch.nil? || batch.empty?

      http = Net::HTTP.new(@uri.host, @uri.port)
      http.use_ssl     = @uri.scheme == 'https'
      http.open_timeout = 3
      http.read_timeout = 5

      req = Net::HTTP::Post.new(@uri.path)
      req['Content-Type'] = 'application/json'
      req['X-Api-Key']    = @api_key
      req.body            = JSON.generate(captures: batch)

      http.request(req)
    rescue
      # fire-and-forget
    end
  end
end
